tests/core/test_configuration.py

```diff
@@ -940,3 +790,474 @@ def test_gettimedelta(self):
+                include_cmds=False,
+            )
+            assert cfg_dict['core'].get('sql_alchemy_conn') == (
+                ('mysql://', "airflow.cfg") if display_source else 'mysql://'
+            )
+            # database should be None because the deprecated value is set in env value
+            assert cfg_dict['database'].get('sql_alchemy_conn') is None
+            if not display_source:
+                remove_all_configurations()
+                conf.read_dict(dictionary=cfg_dict)
+                os.environ.clear()
+                assert conf.get('database', 'sql_alchemy_conn') == 'mysql://'
+
+    @pytest.mark.parametrize("display_source", [True, False])
+    @mock.patch.dict('os.environ', {"AIRFLOW__CORE__SQL_ALCHEMY_CONN": "postgresql://"}, clear=True)
+    def test_conf_as_dict_when_deprecated_value_in_env(self, display_source: bool):
+        with use_config(config="empty.cfg"):
+            cfg_dict = conf.as_dict(
+                display_source=display_source, raw=True, display_sensitive=True, include_env=True
+            )
+            assert cfg_dict['core'].get('sql_alchemy_conn') == (
+                ('postgresql://', "env var") if display_source else 'postgresql://'
+            )
+            # database should be None because the deprecated value is set in env value
+            assert cfg_dict['database'].get('sql_alchemy_conn') is None
+            if not display_source:
+                remove_all_configurations()
+                conf.read_dict(dictionary=cfg_dict)
+                os.environ.clear()
+                assert conf.get('database', 'sql_alchemy_conn') == 'postgresql://'
+
+    @pytest.mark.parametrize("display_source", [True, False])
+    @mock.patch.dict('os.environ', {}, clear=True)
+    def test_conf_as_dict_when_both_conf_and_env_are_empty(self, display_source: bool):
+        with use_config(config="empty.cfg"):
+            cfg_dict = conf.as_dict(display_source=display_source, raw=True, display_sensitive=True)
+            assert cfg_dict['core'].get('sql_alchemy_conn') is None
+            # database should be taken from default because the deprecated value is missing in config
+            assert cfg_dict['database'].get('sql_alchemy_conn') == (
+                (f'sqlite:///{HOME_DIR}/airflow/airflow.db', "default")
+                if display_source
+                else f'sqlite:///{HOME_DIR}/airflow/airflow.db'
+            )
+            if not display_source:
+                remove_all_configurations()
+                conf.read_dict(dictionary=cfg_dict)
+                os.environ.clear()
+                assert conf.get('database', 'sql_alchemy_conn') == f'sqlite:///{HOME_DIR}/airflow/airflow.db'
+
+    @pytest.mark.parametrize("display_source", [True, False])
+    @mock.patch.dict('os.environ', {}, clear=True)
+    def test_conf_as_dict_when_deprecated_value_in_cmd_config(self, display_source: bool):
+        with use_config(config="deprecated_cmd.cfg"):
```

Comment:
Rather than having the tests.utils.test_config.py to replace all configs, do what we do in many of the other tests in this file and create a new AirflowConfigParser -- https://github.com/apache/airflow/blob/6f5749c0d04bd732b320fcbe7713f2611e3d3629/tests/core/test_configuration.py#L241-L249 for example

  That approach also means we don't need a separate file but can inline it directly here which makes it easier (for me at least) to grok the test and look at what it is testing.

I think by making this change it removes the need for the new tests/utils/test_config.py entirely,